### PR TITLE
Improve trace rule processing performance

### PR DESCRIPTION
**Pull request by richardstartin**
- Allows tag to be retrieved from span by name, meaning unmodifiable view unnecessary
- Removes tags from `Rule.processSpan`
- Removes unused traces from `Rule.processSpan`
- Removes duplicated tag lookups
- Inlines "not found" logic into `URLAsResourceNameRule` as this was repeated
- Significantly reduces allocation and CPU usage during scrubbing sensitive info in `URLAsResourceNameRule`

### CODE REVIEW

File: dd-trace-core/src/main/java/datadog/trace/core/processor/rule/URLAsResourceNameRule.java [hunk 1]
```diff
@@ -4,99 +4,164 @@
...
+      return resourceName.toString();
+    } finally {
+      resourceName.setLength(0);
     }
+  }
 
-    final int protoLoc = url.indexOf("://");
-    if (protoLoc < 0) {
-      return url.substring(0, endLoc);
+  private void cleanResourceName(String url, StringBuilder resourceName, int start) {
```
Comment (tylerbenson) [inline dd-trace-core/src/main/java/datadog/trace/core/processor/rule/URLAsResourceNameRule.java:75]: A method comment here describing the algorithm would be helpful.

**Comment by randomanderson** [comment]
One thing that this breaks is the ability to turn off the 404 rule.  Some people might be using `dd.trace.Status404Decorator.enabled=false` .  I think a compromise could adding `Status404Rule` and `Status404Decorator` to the list of aliases

File: dd-trace-core/src/main/java/datadog/trace/core/processor/rule/URLAsResourceNameRule.java [hunk 2]
```diff
@@ -4,99 +4,204 @@
...
-      final String verb = method.toString().toUpperCase().trim();
-      if (!verb.isEmpty()) {
-        path = verb + " " + path;
+    public int indexOf(String text, int from, int to) {
+      int state = 0;
+      to = Math.min(to, text.length());
+      for (int i = from; i < to; ++i) {
+        char c = text.charAt(i);
+        if (c >= 256) { // oops, not latin 1 input
+          state = 0;
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/processor/rule/URLAsResourceNameRule.java:184]: Shouldn't we break out of the loop here?

Comment (richardstartin) [reply dd-trace-core/src/main/java/datadog/trace/core/processor/rule/URLAsResourceNameRule.java:184]: no, unless `to - i < Integer.numberOfTrailingZeros(termination)`, there could still be a match after this character, so we just start again and forget any partial matches so far.

Comment (bantonsson) [reply dd-trace-core/src/main/java/datadog/trace/core/processor/rule/URLAsResourceNameRule.java:184]: Doh. I read this completely backwards. Reminder to self. Never review code on an empty stomach 😉

**Pull request merged**